File: frontend/src/stores/websites.ts
import { ref } from "vue";
import { defineStore } from "pinia";
import api from "@/services/api";

export type RenderingMode = "STATIC" | "DYNAMIC" | "HYBRID";
export type BuildStatus = "idle" | "pending" | "building" | "done" | "failed";
export type XsltSource = "default" | "custom" | "url" | "catalog";
export type XsltProcessor = "lxml" | "saxon";

export interface XsltConfig {
  source: XsltSource;
  content: string | null;
  url: string | null;
  catalog_id: string | null;
  processor: XsltProcessor;
}

export interface AracnePageConfig {
  id: "home" | "browse" | "search";
  sort_order: number;
  is_hidden: boolean;
}

export interface WebsitePage {
  id: string;
  website_id: string;
  slug: string;
  title: string;
  content_md: string | null;
  sort_order: number;
  is_hidden: boolean;
  created_at: string;
  updated_at: string;
}

export interface Website {
  id: string;
  slug: string;
  title: string;
  description: string | null;
  collection_id: string | null;
  rendering_mode: RenderingMode;
  theme_config: Record<string, string>;
  meta_config: Record<string, string | string[]>;
  nav_config: AracnePageConfig[];
  xslt_config: XsltConfig;
  xslt_schema_id: string | null;
  build_status: BuildStatus;
  last_build_at: string | null;
  build_error: string | null;
  is_published: boolean;
  distinct_tags: Record<string, string[]> | null;
  tags_refreshed_at: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  pages: WebsitePage[];
  indices: WebsiteIndex[];
}

export interface WebsiteCreate {
  slug: string;
  title: string;
  description?: string | null;
  collection_id?: string | null;
  rendering_mode?: RenderingMode;
  theme_config?: Record<string, string>;
  meta_config?: Record<string, string | string[]>;
  is_published?: boolean;
}

export interface WebsiteUpdate {
  title?: string;
  description?: string | null;
  collection_id?: string | null;
  rendering_mode?: RenderingMode;
  theme_config?: Record<string, string>;
  meta_config?: Record<string, string | string[]>;
  nav_config?: AracnePageConfig[];
  xslt_config?: XsltConfig;
  is_published?: boolean;
}

export interface WebsitePageCreate {
  slug: string;
  title: string;
  content_md?: string | null;
  sort_order?: number;
  is_hidden?: boolean;
}

export interface WebsitePageUpdate {
  title?: string;
  content_md?: string | null;
  sort_order?: number;
  is_hidden?: boolean;
}

export interface WebsiteIndex {
  id: string;
  website_id: string;
  label: string;
  title: string;
  tag: string;
  key_attribute: string | null;
  subkey_attribute: string | null;
  last_built_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface WebsiteIndexCreate {
  label: string;
  title: string;
  tag: string;
  key_attribute?: string | null;
  subkey_attribute?: string | null;
}

export interface WebsiteIndexUpdate {
  label?: string;
  title?: string;
  tag?: string;
  key_attribute?: string | null;
  subkey_attribute?: string | null;
}

export interface MetaSuggestions {
  author: string[];
  dc_creator: string[];
  designer: string[];
  copyright: string;
  dc_publisher: string[];
  dc_format: string;
  dc_identifier: string;
}

export const useWebsiteStore = defineStore("websites", () => {
  const websites = ref<Website[]>([]);
  const isLoading = ref(false);
  const error = ref<string | null>(null);

  async function fetchWebsites(): Promise<void> {
    isLoading.value = true;
    error.value = null;
    try {
      const res = await api.get<Website[]>("/websites");
      websites.value = res.data.data as Website[];
    } catch (err: unknown) {
      error.value = err instanceof Error ? err.message : "Error loading websites";
    } finally {
      isLoading.value = false;
    }
  }

  async function createWebsite(data: WebsiteCreate): Promise<Website> {
    const res = await api.post<Website>("/websites", data);
    const created = res.data.data as Website;
    websites.value.unshift(created);
    return created;
  }

  async function updateWebsite(slug: string, data: WebsiteUpdate): Promise<Website> {
    const res = await api.put<Website>(`/websites/${slug}`, data);
    const updated = res.data.data as Website;
    const idx = websites.value.findIndex((w) => w.slug === slug);
    if (idx !== -1) websites.value[idx] = updated;
    return updated;
  }

  async function deleteWebsite(slug: string): Promise<void> {
    await api.delete(`/websites/${slug}`);
    websites.value = websites.value.filter((w) => w.slug !== slug);
  }

  async function triggerBuild(slug: string): Promise<void> {
    await api.post(`/websites/${slug}/build`);
    // Refresh to get updated build_status
    const res = await api.get<Website>(`/websites/${slug}`);
    const updated = res.data.data as Website;
    const idx = websites.value.findIndex((w) => w.slug === slug);
    if (idx !== -1) websites.value[idx] = updated;
  }

  async function pollBuildStatus(slug: string): Promise<BuildStatus> {
    const res = await api.get<Website>(`/websites/${slug}`);
    const updated = res.data.data as Website;
    const idx = websites.value.findIndex((w) => w.slug === slug);
    if (idx !== -1) websites.value[idx] = updated;
    return updated.build_status;
  }

  async function createPage(slug: string, data: WebsitePageCreate): Promise<WebsitePage> {
    const res = await api.post<WebsitePage>(`/websites/${slug}/pages`, data);
    const page = res.data.data as WebsitePage;
    const site = websites.value.find((w) => w.slug === slug);
    if (site) site.pages.push(page);
    return page;
  }

  async function updatePage(
    slug: string,
    pageSlug: string,
    data: WebsitePageUpdate,
  ): Promise<WebsitePage> {
    const res = await api.put<WebsitePage>(`/websites/${slug}/pages/${pageSlug}`, data);
    const updated = res.data.data as WebsitePage;
    const site = websites.value.find((w) => w.slug === slug);
    if (site) {
      const idx = site.pages.findIndex((p) => p.slug === pageSlug);
      if (idx !== -1) site.pages[idx] = updated;
    }
    return updated;
  }

  async function deletePage(slug: string, pageSlug: string): Promise<void> {
    await api.delete(`/websites/${slug}/pages/${pageSlug}`);
    const site = websites.value.find((w) => w.slug === slug);
    if (site) site.pages = site.pages.filter((p) => p.slug !== pageSlug);
  }

  async function fetchMetaSuggestions(slug: string): Promise<MetaSuggestions> {
    const res = await api.get<MetaSuggestions>(`/websites/${slug}/meta-suggestions`);
    return res.data.data as MetaSuggestions;
  }

  /**
   * Apply the given XSLT config to a single document and return the body HTML.
   *
   * Pass xsltConfig to preview unsaved stylesheet changes; omit it (or pass
   * null) to use the website's currently saved xslt_config.
   */
  async function previewDocument(
    slug: string,
    filename: string,
    xsltConfig?: XsltConfig | null,
  ): Promise<string> {
    const res = await api.post<{ html: string }>(
      `/websites/${slug}/preview-doc/${encodeURIComponent(filename)}`,
      { xslt_config: xsltConfig ?? null },
    );
    return (res.data.data as { html: string }).html;
  }

  /** Invalidate the in-memory rendered-page and XSLT caches for *slug*. */
  async function clearCache(slug: string): Promise<void> {
    await api.post(`/websites/${slug}/clear-cache`, {});
  }

  /** Download the built STATIC site as a ZIP file. */
  async function downloadSite(slug: string): Promise<void> {
    const res = await api.get(`/websites/${slug}/download`, { responseType: "blob" });
    const url = URL.createObjectURL(res.data as Blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `${slug}.zip`;
    a.click();
    URL.revokeObjectURL(url);
  }

  // ── Tag discovery ────────────────────────────────────────────────────────

  async function refreshTags(slug: string): Promise<{ distinct_tags: Record<string, string[]> | null; tags_refreshed_at: string | null }> {
    const res = await api.post(`/websites/${slug}/tags/refresh`, {});
    const data = res.data.data as { distinct_tags: Record<string, string[]> | null; tags_refreshed_at: string | null };
    const site = websites.value.find((w) => w.slug === slug);
    if (site) {
      site.distinct_tags = data.distinct_tags;
      site.tags_refreshed_at = data.tags_refreshed_at;
    }
    return data;
  }

  // ── Website indices ──────────────────────────────────────────────────────

  async function createIndex(slug: string, data: WebsiteIndexCreate): Promise<WebsiteIndex> {
    const res = await api.post<WebsiteIndex>(`/websites/${slug}/indices`, data);
    const created = res.data.data as WebsiteIndex;
    const site = websites.value.find((w) => w.slug === slug);
    if (site) site.indices.push(created);
    return created;
  }

  async function updateIndex(slug: string, indexId: string, data: WebsiteIndexUpdate): Promise<WebsiteIndex> {
    const res = await api.put<WebsiteIndex>(`/websites/${slug}/indices/${indexId}`, data);
    const updated = res.data.data as WebsiteIndex;
    const site = websites.value.find((w) => w.slug === slug);
    if (site) {
      const i = site.indices.findIndex((x) => x.id === indexId);
      if (i !== -1) site.indices[i] = updated;
    }
    return updated;
  }

  async function deleteIndex(slug: string, indexId: string): Promise<void> {
    await api.delete(`/websites/${slug}/indices/${indexId}`);
    const site = websites.value.find((w) => w.slug === slug);
    if (site) site.indices = site.indices.filter((x) => x.id !== indexId);
  }

  async function rebuildIndex(slug: string, indexId: string): Promise<WebsiteIndex> {
    const res = await api.post<WebsiteIndex>(`/websites/${slug}/indices/${indexId}/rebuild`, {});
    const updated = res.data.data as WebsiteIndex;
    const site = websites.value.find((w) => w.slug === slug);
    if (site) {
      const i = site.indices.findIndex((x) => x.id === indexId);
      if (i !== -1) site.indices[i] = updated;
    }
    return updated;
  }

  async function rebuildAllIndices(slug: string): Promise<WebsiteIndex[]> {
    const res = await api.post<WebsiteIndex[]>(`/websites/${slug}/indices/rebuild-all`, {});
    const updated = res.data.data as WebsiteIndex[];
    const site = websites.value.find((w) => w.slug === slug);
    if (site) site.indices = updated;
    return updated;
  }

  return {
    websites,
    isLoading,
    error,
    fetchWebsites,
    createWebsite,
    updateWebsite,
    deleteWebsite,
    triggerBuild,
    pollBuildStatus,
    createPage,
    updatePage,
    deletePage,
    fetchMetaSuggestions,
    previewDocument,
    clearCache,
    downloadSite,
    refreshTags,
    createIndex,
    updateIndex,
    deleteIndex,
    rebuildIndex,
    rebuildAllIndices,
  };
});

File: backend/app/schemas/websites.py
"""Pydantic v2 schemas for websites, website pages, and website indices."""

import uuid
from datetime import datetime
from urllib.parse import urlparse

from pydantic import BaseModel, Field, field_validator

from app.models.website import BuildStatus, RenderingMode


# ── Website pages ─────────────────────────────────────────────────────────────

class WebsitePageCreate(BaseModel):
    slug: str = Field(..., min_length=1, max_length=128, pattern=r"^[a-z0-9_-]+$")
    title: str = Field(..., min_length=1, max_length=256)
    content_md: str | None = None
    sort_order: int = 0
    is_hidden: bool = False


class WebsitePageUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=256)
    content_md: str | None = None
    sort_order: int | None = None
    is_hidden: bool | None = None


class WebsitePageResponse(BaseModel):
    id: uuid.UUID
    website_id: uuid.UUID
    slug: str
    title: str
    content_md: str | None
    sort_order: int
    is_hidden: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Websites ──────────────────────────────────────────────────────────────────

def _validate_url_scheme(v: str | None) -> str | None:
    """Reject non-http/https schemes to prevent javascript: stored XSS."""
    if not v:
        return v
    v = v.strip()
    parsed = urlparse(v)
    if parsed.scheme not in ("http", "https"):
        raise ValueError("website_url must start with http:// or https://")
    if not parsed.netloc:
        raise ValueError("website_url must include a valid hostname")
    return v


class WebsiteCreate(BaseModel):
    slug: str = Field(..., min_length=1, max_length=128, pattern=r"^[a-z0-9_-]+$")
    title: str = Field(..., min_length=1, max_length=256)
    description: str | None = None
    collection_id: uuid.UUID | None = None
    rendering_mode: RenderingMode = RenderingMode.STATIC
    website_url: str | None = Field(None, max_length=512)
    theme_config: dict = Field(default_factory=dict)
    meta_config: dict = Field(default_factory=dict)
    nav_config: list = Field(default_factory=list)
    xslt_config: dict = Field(default_factory=dict)
    xslt_schema_id: uuid.UUID | None = None
    is_published: bool = False
    show_in_public_home: bool = False
    custom_css: str | None = None
    custom_js: str | None = None
    include_jquery: bool = False
    maintenance_on_unpublish: bool | None = None  # None → set per-mode default in service
    maintenance_message: str | None = Field(None, max_length=1024)
    contact_email: str | None = Field(None, max_length=256)

    @field_validator("website_url", mode="before")
    @classmethod
    def validate_website_url(cls, v: str | None) -> str | None:
        return _validate_url_scheme(v)


class WebsiteUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=256)
    description: str | None = None
    collection_id: uuid.UUID | None = None
    rendering_mode: RenderingMode | None = None
    website_url: str | None = Field(None, max_length=512)

    @field_validator("website_url", mode="before")
    @classmethod
    def validate_website_url(cls, v: str | None) -> str | None:
        return _validate_url_scheme(v)
    theme_config: dict | None = None
    meta_config: dict | None = None
    nav_config: list | None = None
    xslt_config: dict | None = None
    xslt_schema_id: uuid.UUID | None = None
    is_published: bool | None = None
    show_in_public_home: bool | None = None
    custom_css: str | None = None
    custom_js: str | None = None
    include_jquery: bool | None = None
    maintenance_on_unpublish: bool | None = None
    maintenance_message: str | None = Field(None, max_length=1024)
    contact_email: str | None = Field(None, max_length=256)


# ── Website indices ───────────────────────────────────────────────────────────

class WebsiteIndexCreate(BaseModel):
    label: str = Field(..., min_length=1, max_length=128, pattern=r"^[a-z0-9_-]+$")
    title: str = Field(..., min_length=1, max_length=256)
    tag: str = Field(..., min_length=1, max_length=128)
    key_attribute: str | None = Field(None, max_length=128)
    subkey_attribute: str | None = Field(None, max_length=128)


class WebsiteIndexUpdate(BaseModel):
    label: str | None = Field(None, min_length=1, max_length=128, pattern=r"^[a-z0-9_-]+$")
    title: str | None = Field(None, min_length=1, max_length=256)
    tag: str | None = Field(None, min_length=1, max_length=128)
    key_attribute: str | None = None
    subkey_attribute: str | None = None


class WebsiteIndexResponse(BaseModel):
    id: uuid.UUID
    website_id: uuid.UUID
    label: str
    title: str
    tag: str
    key_attribute: str | None
    subkey_attribute: str | None
    last_built_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Websites ──────────────────────────────────────────────────────────────────

class WebsiteResponse(BaseModel):
    id: uuid.UUID
    slug: str
    title: str
    description: str | None
    collection_id: uuid.UUID | None
    rendering_mode: RenderingMode
    website_url: str | None
    theme_config: dict
    meta_config: dict
    nav_config: list
    xslt_config: dict
    xslt_schema_id: uuid.UUID | None
    build_status: BuildStatus
    last_build_at: datetime | None
    build_error: str | None
    is_published: bool
    show_in_public_home: bool
    custom_css: str | None
    custom_js: str | None
    include_jquery: bool
    maintenance_on_unpublish: bool
    maintenance_message: str | None
    contact_email: str | None
    distinct_tags: dict | list | None  # dict when populated, [] when XQuery returned no results
    tags_refreshed_at: datetime | None
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime
    pages: list[WebsitePageResponse] = Field(default_factory=list)
    indices: list[WebsiteIndexResponse] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class WebsiteBuildResponse(BaseModel):
    """Returned immediately when a build is triggered."""
    slug: str
    build_status: BuildStatus
    message: str


class MetaSuggestionsResponse(BaseModel):
    """Pre-computed meta field suggestions for the edit form."""
    author: list[str] = []
    dc_creator: list[str] = []
    designer: list[str] = []
    copyright: str = ""
    dc_publisher: list[str] = []
    dc_format: str = ""
    dc_identifier: str = ""


# ── XSLT preview ──────────────────────────────────────────────────────────────

class WebsitePreviewDocRequest(BaseModel):
    """Optional XSLT config override for the preview-doc endpoint.

    If xslt_config is omitted or null, the endpoint uses the website's saved
    xslt_config.  Pass a full XsltConfig dict to preview unsaved changes.
    """
    xslt_config: dict | None = None


class WebsitePreviewDocResponse(BaseModel):
    """HTML body returned by the preview-doc endpoint."""
    html: str


# ── Cache management ──────────────────────────────────────────────────────────

class WebsiteCacheClearedResponse(BaseModel):
    """Returned by the clear-cache endpoint."""
    cleared: bool


# ── Tags ──────────────────────────────────────────────────────────────────────

class WebsiteTagsResponse(BaseModel):
    """Distinct-tag map for a website's linked collection.

    distinct_tags may be a dict (element → values map), an empty list
    (XQuery returned no results), or None (no refresh has been run yet).
    """
    distinct_tags: dict | list | None
    tags_refreshed_at: datetime | None
